File: packages/core/test/crypto.test.ts
import { createHmac } from "node:crypto";
import { describe, expect, it } from "vitest";
import { sha256Hex, verifyGithubSignature } from "../src/crypto.js";

describe("verifyGithubSignature", () => {
  const secret = "s3cret";
  const body = JSON.stringify({ hello: "world" });
  const goodSig = "sha256=" + createHmac("sha256", secret).update(body).digest("hex");

  it("accepts a valid signature", () => {
    expect(verifyGithubSignature(secret, body, goodSig)).toBe(true);
  });

  it("rejects a tampered body", () => {
    expect(verifyGithubSignature(secret, body + " ", goodSig)).toBe(false);
  });

  it("rejects the wrong secret", () => {
    const sig = "sha256=" + createHmac("sha256", "other").update(body).digest("hex");
    expect(verifyGithubSignature(secret, body, sig)).toBe(false);
  });

  it("rejects missing/malformed headers", () => {
    expect(verifyGithubSignature(secret, body, undefined)).toBe(false);
    expect(verifyGithubSignature(secret, body, "sha1=abc")).toBe(false);
    expect(verifyGithubSignature(secret, body, "sha256=zz")).toBe(false);
  });
});

describe("sha256Hex", () => {
  it("hashes deterministically", () => {
    expect(sha256Hex("token")).toBe(sha256Hex("token"));
    expect(sha256Hex("token")).toHaveLength(64);
  });
});

File: internal/httpapi/crypto_test.go
package httpapi

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"testing"
)

func sign(secret, body string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

func TestVerifyGithubSignature(t *testing.T) {
	secret := "s3cret"
	body := `{"hello":"world"}`
	good := sign(secret, body)

	if !VerifyGithubSignature(secret, []byte(body), good) {
		t.Error("valid signature rejected")
	}
	if VerifyGithubSignature(secret, []byte(body+" "), good) {
		t.Error("tampered body accepted")
	}
	if VerifyGithubSignature(secret, []byte(body), sign("other", body)) {
		t.Error("wrong secret accepted")
	}
	if VerifyGithubSignature(secret, []byte(body), "") {
		t.Error("missing header accepted")
	}
	if VerifyGithubSignature(secret, []byte(body), "sha1=abc") {
		t.Error("sha1 header accepted")
	}
	if VerifyGithubSignature(secret, []byte(body), "sha256=zz") {
		t.Error("malformed hex accepted")
	}
}

func TestSha256Hex(t *testing.T) {
	if Sha256Hex("token") != Sha256Hex("token") {
		t.Error("not deterministic")
	}
	if len(Sha256Hex("token")) != 64 {
		t.Error("wrong length")
	}
}
